diff --git a/src/ViennaRNA/constraints/soft.c b/src/ViennaRNA/constraints/soft.c
index dfc5c0d..c89bcc9 100644
--- a/src/ViennaRNA/constraints/soft.c
+++ b/src/ViennaRNA/constraints/soft.c
@@ -176,10 +176,10 @@ sc_store_bp(vrna_sc_bp_storage_t  **container,
             unsigned int          i,
             unsigned int          start,
             unsigned int          end,
-            int                   e);
+            FLT_OR_DBL            e);
 
 
-PRIVATE INLINE int
+PRIVATE INLINE FLT_OR_DBL
 get_stored_bp_contributions(vrna_sc_bp_storage_t  *container,
                             unsigned int          j);
 
@@ -1344,10 +1344,9 @@ PRIVATE INLINE void
 sc_init_up_storage(vrna_sc_t *sc)
 {
   if (!sc->up_storage)
-    sc->up_storage = (int *)vrna_alloc(sizeof(int) * (sc->n + 2));
+    sc->up_storage = (double *)vrna_alloc(sizeof(double) * (sc->n + 2));
 }
 
-
 /* pupulate sc->energy_up array at position i from sc->up_storage data */
 PRIVATE INLINE void
 populate_sc_up_mfe(vrna_sc_t    *sc,
@@ -1359,7 +1358,7 @@ populate_sc_up_mfe(vrna_sc_t    *sc,
   sc->energy_up[i][0] = 0;
   for (j = 1; j <= n; j++)
     sc->energy_up[i][j] = sc->energy_up[i][j - 1] +
-                          sc->up_storage[i + j - 1];
+                          (int)roundf(sc->up_storage[i + j - 1]);
 }
 
 
@@ -1403,7 +1402,7 @@ sc_store_bp(vrna_sc_bp_storage_t  **container,
             unsigned int          i,
             unsigned int          start,
             unsigned int          end,
-            int                   e)
+            FLT_OR_DBL            e)
 {
   unsigned int size, cnt = 0;
 
@@ -1435,14 +1434,14 @@ sc_store_bp(vrna_sc_bp_storage_t  **container,
 }
 
 
-PRIVATE INLINE int
+PRIVATE INLINE FLT_OR_DBL
 get_stored_bp_contributions(vrna_sc_bp_storage_t  *container,
                             unsigned int          j)
 {
   unsigned int  cnt;
-  int           e;
+  FLT_OR_DBL e;
 
-  e = 0;
+  e = 0.0;
 
   /* go through list of constraints for current position i */
   for (cnt = 0; container[cnt].interval_start != 0; cnt++) {
@@ -1468,7 +1467,7 @@ populate_sc_bp_mfe(vrna_sc_t    *sc,
                    int          *idx)
 {
   unsigned int  j, k;
-  int           e;
+  FLT_OR_DBL       e;
 
   if (sc->bp_storage[i]) {
     for (k = 1; k < maxdist; k++) {
@@ -1481,11 +1480,11 @@ populate_sc_bp_mfe(vrna_sc_t    *sc,
 
       switch (sc->type) {
         case VRNA_SC_DEFAULT:
-          sc->energy_bp[idx[j] + i] = e;
+          sc->energy_bp[idx[j] + i] = (int)roundf(e);
           break;
 
         case VRNA_SC_WINDOW:
-          sc->energy_bp_local[i][j - i] = e;
+          sc->energy_bp_local[i][j - i] = (int)roundf(e);
           break;
       }
     }
@@ -1518,7 +1517,7 @@ populate_sc_bp_pf(vrna_sc_t     *sc,
                   double        kT)
 {
   unsigned int      j, k;
-  int               e;
+  FLT_OR_DBL        e;
   FLT_OR_DBL        q;
   double            GT;
 
@@ -1580,7 +1579,7 @@ sc_add_bp(vrna_fold_compound_t  *fc,
 
   sc = fc->sc;
   sc_init_bp_storage(sc);
-  sc_store_bp(sc->bp_storage, i, j, j, (int)roundf(energy * 100.));
+  sc_store_bp(sc->bp_storage, i, j, j, energy * 100.);
   sc->state |= STATE_DIRTY_BP_MFE | STATE_DIRTY_BP_PF;
 }
 
@@ -1712,7 +1711,7 @@ sc_reset_up(vrna_fold_compound_t  *fc,
 
     /* add contributions to storage container */
     for (i = 1; i <= n; i++)
-      sc->up_storage[i] = (int)roundf(constraints[i] * 100.); /* convert to 10kal/mol */
+      sc->up_storage[i] = constraints[i] * 100.; /* convert to 10kal/mol */
 
     sc->state |= STATE_DIRTY_UP_MFE | STATE_DIRTY_UP_PF;
   } else {
@@ -1752,7 +1751,7 @@ sc_reset_up_comparative(vrna_fold_compound_t  *fc,
 
       /* add contributions to storage container */
       for (i = 1; i <= n; i++)
-        sc->up_storage[i] = (int)roundf(constraints[s][i] * 100.); /* convert to 10kal/mol */
+        sc->up_storage[i] = constraints[s][i] * 100.; /* convert to 10kal/mol */
 
       sc->state |= STATE_DIRTY_UP_MFE | STATE_DIRTY_UP_PF;
 
@@ -1794,7 +1793,7 @@ sc_reset_bp(vrna_fold_compound_t  *fc,
     /* add contributions to storage container */
     for (i = 1; i < n; i++)
       for (j = i + 1; j <= n; j++)
-        sc_store_bp(sc->bp_storage, i, j, j, (int)roundf(constraints[i][j] * 100.));
+        sc_store_bp(sc->bp_storage, i, j, j, constraints[i][j] * 100.);
 
     sc->state |= STATE_DIRTY_BP_MFE | STATE_DIRTY_BP_PF;
   } else {
@@ -1835,7 +1834,7 @@ sc_reset_bp_comparative(vrna_fold_compound_t  *fc,
       /* add contributions to storage container */
       for (i = 1; i < n; i++)
         for (j = i + 1; j <= n; j++)
-          sc_store_bp(sc->bp_storage, i, j, j, (int)roundf(constraints[s][i][j] * 100.));
+          sc_store_bp(sc->bp_storage, i, j, j, constraints[s][i][j] * 100.);
 
       sc->state |= STATE_DIRTY_BP_MFE | STATE_DIRTY_BP_PF;
 
@@ -1864,7 +1863,7 @@ sc_add_up(vrna_fold_compound_t  *fc,
 
   sc = fc->sc;
   sc_init_up_storage(sc);
-  sc->up_storage[i] += (int)roundf(energy * 100.);
+  sc->up_storage[i] += energy * 100.;
   sc->state         |= STATE_DIRTY_UP_MFE | STATE_DIRTY_UP_PF;
 }
 
@@ -1892,7 +1891,7 @@ sc_add_up_comparative(vrna_fold_compound_t  *fc,
     if (is[s] != 0) {
       sc = fc->scs[s];
       sc_init_up_storage(sc);
-      sc->up_storage[is[s]] += (int)roundf(energies[s] * 100.);
+      sc->up_storage[is[s]] += energies[s] * 100.;
       sc->state             |= STATE_DIRTY_UP_MFE | STATE_DIRTY_UP_PF;
 
       ret++;
diff --git a/src/ViennaRNA/constraints/soft.h b/src/ViennaRNA/constraints/soft.h
index 19a5e90..f4471b4 100644
--- a/src/ViennaRNA/constraints/soft.h
+++ b/src/ViennaRNA/constraints/soft.h
@@ -221,7 +221,7 @@ typedef enum {
 typedef struct {
   unsigned int  interval_start;
   unsigned int  interval_end;
-  int           e;
+  FLT_OR_DBL   	e;
 } vrna_sc_bp_storage_t;
 
 
@@ -243,7 +243,7 @@ struct vrna_sc_s {
   int                   **energy_up;      /**<  @brief Energy contribution for stretches of unpaired nucleotides */
   FLT_OR_DBL            **exp_energy_up;  /**<  @brief Boltzmann Factors of the energy contributions for unpaired sequence stretches */
 
-  int                   *up_storage;      /**<  @brief  Storage container for energy contributions per unpaired nucleotide */
+  FLT_OR_DBL            *up_storage;      /**<  @brief  Storage container for energy contributions per unpaired nucleotide */
   vrna_sc_bp_storage_t  **bp_storage;     /**<  @brief  Storage container for energy contributions per base pair */
 
   int           *energy_stack;                    /**<  @brief Pseudo Energy contribution per base pair involved in a stack */
